File: android/src/main/kotlin/com/perkox/perkox_flutter_sdk/PerkoxFlutterSdkPlugin.kt
package com.perkox.perkox_flutter_sdk

import android.app.Activity
import android.os.Handler
import android.os.Looper
import com.perkoxofferwall.sdk.Offerwall
import com.perkoxofferwall.sdk.PerkoxOfferwall
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * PerkoxFlutterSdkPlugin
 * Bridges Flutter calls to native Perkox Android SDK.
 */
class PerkoxFlutterSdkPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var currentActivity: Activity? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private var activeAppId: String = ""
    private var activeSdkKey: String = ""
    private var activePlayerId: String = ""
    private var activeBeta: Boolean = false

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initSDK" -> {
                val appId = call.argument<String>("appId") ?: ""
                val sdkKey = call.argument<String>("sdkKey") ?: ""
                val playerId = call.argument<String>("playerId") ?: ""
                val beta = call.argument<Boolean>("beta") ?: false

                activeAppId = if (appId != "undefined") appId else ""
                activeSdkKey = if (sdkKey != "undefined") sdkKey else ""
                activePlayerId = if (playerId != "undefined") playerId else ""
                activeBeta = beta

                result.success(true)
            }

            "setUserId" -> {
                val playerId = call.argument<String>("playerId") ?: ""
                activePlayerId = if (playerId != "undefined") playerId else ""
                result.success(true)
            }

            "showOfferwall" -> {
                val activity = currentActivity
                if (activity == null) {
                    result.error("NO_ACTIVITY", "Cannot show offerwall without an active foreground Activity", null)
                    return
                }

                try {
                    val appIdArg = call.argument<String>("appId")
                    val sdkKeyArg = call.argument<String>("sdkKey")
                    val playerIdArg = call.argument<String>("playerId")
                    val betaArg = call.argument<Boolean>("beta") ?: activeBeta

                    val app = if (!appIdArg.isNullOrEmpty() && appIdArg != "undefined") appIdArg else activeAppId
                    val key = if (!sdkKeyArg.isNullOrEmpty() && sdkKeyArg != "undefined") sdkKeyArg else activeSdkKey
                    val user = if (!playerIdArg.isNullOrEmpty() && playerIdArg != "undefined") playerIdArg else activePlayerId

                    if (app.isEmpty() || key.isEmpty()) {
                        result.error("INVALID_CONFIG", "appId and sdkKey must be provided", null)
                        return
                    }

                    val offerwall: Offerwall = PerkoxOfferwall.create(app, key, user)

                    offerwall.onReward = { rewardMap ->
                        mainHandler.post {
                            val data = rewardMap ?: emptyMap<String, Any?>()
                            // Dispatch via MethodChannel invokeMethod
                            methodChannel.invokeMethod("onPerkoxReward", data)

                            // Also dispatch via EventChannel
                            eventSink?.success(
                                mapOf(
                                    "event" to "onPerkoxReward",
                                    "data" to data
                                )
                            )
                        }
                    }

                    offerwall.onClose = {
                        mainHandler.post {
                            // Dispatch via MethodChannel invokeMethod
                            methodChannel.invokeMethod("onPerkoxClose", null)

                            // Also dispatch via EventChannel
                            eventSink?.success(
                                mapOf(
                                    "event" to "onPerkoxClose"
                                )
                            )
                        }
                    }

                    offerwall.launch(activity, betaArg)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("PERKOX_LAUNCH_ERROR", e.message, e.stackTraceToString())
                }
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // --- ActivityAware Lifecycle ---
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        currentActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
    }

    override fun onDetachedFromActivity() {
        currentActivity = null
    }

    // --- StreamHandler for EventChannel ---
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    companion object {
        const val METHOD_CHANNEL_NAME = "com.perkox.flutter_sdk/methods"
        const val EVENT_CHANNEL_NAME = "com.perkox.flutter_sdk/events"
    }
}

File: ios/perkox_flutter_sdk/Sources/perkox_flutter_sdk/PerkoxFlutterSdkPlugin.swift
import Flutter
import UIKit

#if canImport(PerkoxOfferwall)
import PerkoxOfferwall
#endif

public class PerkoxFlutterSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var activeAppId: String = ""
    private var activeSdkKey: String = ""
    private var activePlayerId: String = ""
    private var activeBeta: Bool = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.perkox.flutter_sdk/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.perkox.flutter_sdk/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = PerkoxFlutterSdkPlugin()
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initSDK":
            if let args = call.arguments as? [String: Any] {
                let appId = args["appId"] as? String ?? ""
                let sdkKey = args["sdkKey"] as? String ?? ""
                let playerId = args["playerId"] as? String ?? ""
                let beta = args["beta"] as? Bool ?? false

                activeAppId = appId != "undefined" ? appId : ""
                activeSdkKey = sdkKey != "undefined" ? sdkKey : ""
                activePlayerId = playerId != "undefined" ? playerId : ""
                activeBeta = beta
            }
            result(true)

        case "setUserId":
            if let args = call.arguments as? [String: Any] {
                let playerId = args["playerId"] as? String ?? ""
                activePlayerId = playerId != "undefined" ? playerId : ""
            }
            result(true)

        case "showOfferwall":
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }

                guard let topVC = self.getTopViewController() else {
                    result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "Unable to find top view controller to present offerwall", details: nil))
                    return
                }

                let args = call.arguments as? [String: Any]
                let appIdArg = args?["appId"] as? String
                let sdkKeyArg = args?["sdkKey"] as? String
                let playerIdArg = args?["playerId"] as? String
                let betaArg = args?["beta"] as? Bool ?? self.activeBeta

                let app = (appIdArg != nil && !appIdArg!.isEmpty && appIdArg != "undefined") ? appIdArg! : self.activeAppId
                let key = (sdkKeyArg != nil && !sdkKeyArg!.isEmpty && sdkKeyArg != "undefined") ? sdkKeyArg! : self.activeSdkKey
                let user = (playerIdArg != nil && !playerIdArg!.isEmpty && playerIdArg != "undefined") ? playerIdArg! : self.activePlayerId

                if app.isEmpty || key.isEmpty {
                    result(FlutterError(code: "INVALID_CONFIG", message: "appId and sdkKey must be provided", details: nil))
                    return
                }

                #if canImport(PerkoxOfferwall)
                let offerwall = PerkoxOfferwall.create(
                    appId: app,
                    sdkKey: key,
                    playerId: user
                )

                offerwall.onReward = { [weak self] reward in
                    var rewardMap: [String: Any] = [:]
                    for (k, v) in reward {
                        if let val = v {
                            rewardMap[k] = val
                        }
                    }

                    DispatchQueue.main.async {
                        // Dispatch via MethodChannel
                        self?.methodChannel?.invokeMethod("onPerkoxReward", arguments: rewardMap)

                        // Dispatch via EventChannel
                        self?.eventSink?([
                            "event": "onPerkoxReward",
                            "data": rewardMap
                        ])
                    }
                }

                offerwall.onClose = { [weak self] in
                    DispatchQueue.main.async {
                        // Dispatch via MethodChannel
                        self?.methodChannel?.invokeMethod("onPerkoxClose", arguments: nil)

                        // Dispatch via EventChannel
                        self?.eventSink?([
                            "event": "onPerkoxClose"
                        ])
                    }
                }

                offerwall.launch(viewController: topVC, beta: betaArg)
                result(true)
                #else
                result(FlutterError(code: "PERKOX_IOS_SDK_MISSING", message: "PerkoxOfferwall.xcframework is not linked in podspec", details: nil))
                #endif
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getTopViewController(from rootVC: UIViewController? = nil) -> UIViewController? {
        let root = rootVC ?? {
            if #available(iOS 13.0, *) {
                return UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }
                    .flatMap { $0.windows }
                    .first(where: { $0.isKeyWindow })?.rootViewController
            } else {
                return UIApplication.shared.keyWindow?.rootViewController
            }
        }()

        if let nav = root as? UINavigationController {
            return getTopViewController(from: nav.visibleViewController)
        }
        if let tab = root as? UITabBarController {
            return getTopViewController(from: tab.selectedViewController)
        }
        if let presented = root?.presentedViewController {
            return getTopViewController(from: presented)
        }
        return root
    }

    // --- FlutterStreamHandler ---
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
